spring-boot/admin/src/test/java/eu/coatrack/admin/service/GatewayHealthMonitorServiceTest.java

```diff
@@ -0,0 +1,96 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.mockito.Mockito.when;
+
+public class GatewayHealthMonitorServiceTest {
+
+    @Mock
+    ProxyRepository proxyRepository;
+
+    @Mock
+    Proxy proxySample;
+
+    private List<Proxy> sampleProxies = new ArrayList<>();
+
+    @InjectMocks
+    private GatewayHealthMonitorService gatewayHealthMonitorService;
+
+    public Proxy createSampleProxy(boolean monitoredEnabled, String proxyName) {
+        Proxy proxy = new Proxy();
+        proxy.setName(proxyName);
+        proxy.setId("testIdFor-" + proxyName);
+        proxy.setMonitoringEnabled(monitoredEnabled);
+        proxy.updateTimeOfLastSuccessfulCallToAdmin_setToNow();
+        return proxy;
+    }
+
+    @Before
+    public void setup() {
+        MockitoAnnotations.initMocks(this);
+        sampleProxies = Arrays.asList(createSampleProxy(true, "testProxyMonitoringEnabledA"), createSampleProxy(false, "testProxyMonitoringDisabledA"), proxySample);
+        when(proxyRepository.findAvailable()).thenReturn(sampleProxies);
+        when(proxySample.isMonitoringEnabled()).thenReturn(true);
+        when(proxySample.getName()).thenReturn("testProxyMonitoringEnabledB");
+        ReflectionTestUtils.setField(gatewayHealthMonitorService, "gatewayHealthWarningThresholdInMinutes", 5);
+        ReflectionTestUtils.setField(gatewayHealthMonitorService, "gatewayHealthCriticalThresholdInMinutes", 60);
+    }
+
+    @Test
+    public void ifAllGatewaysAreOkOrIgnore_ThenStatusSummaryShouldReturnOkState() {
+        when(proxySample.getTimeOfLastSuccessfulCallToAdmin()).thenReturn(LocalDateTime.now());
+        Assert.assertEquals(ProxyStates.OK,
+                gatewayHealthMonitorService
+                        .calculateGatewayHealthStatusSummary(gatewayHealthMonitorService.getGatewayHealthMonitorData()));
+    }
+
+    @Test
+    public void ifOneGatewayIsWarningAndOthersOKOrIgnore_ThenStatusSummaryShouldReturnWarningState() {
+        when(proxySample.getTimeOfLastSuccessfulCallToAdmin()).thenReturn(LocalDateTime.now().minusMinutes(6));
+        Assert.assertEquals(ProxyStates.WARNING,
+                gatewayHealthMonitorService.
+                        calculateGatewayHealthStatusSummary(gatewayHealthMonitorService.getGatewayHealthMonitorData()));
+    }
+
+    @Test
+    public void ifOneGatewayIsCritical_ThenStatusSummaryShouldReturnCriticalState() {
+        when(proxySample.getTimeOfLastSuccessfulCallToAdmin()).thenReturn(LocalDateTime.now().minusMinutes(70));
+        Assert.assertEquals(ProxyStates.CRITICAL,
+                gatewayHealthMonitorService
+                        .calculateGatewayHealthStatusSummary(gatewayHealthMonitorService.getGatewayHealthMonitorData()));
+    }
+
+    @Test
+    public void ifAllGatewaysNotConnected_ThenStatusSummaryShouldReturnNotConnectedState() {
+        when(proxySample.getTimeOfLastSuccessfulCallToAdmin()).thenReturn(null);
```

Comment:
I would assume that making the entity field `Optional` would lead to problems in object-relational mapping when storing in the database. However, the getter could obviously wrap the nullable database attribute inside an `Optional`.

Reply:
Then we should keep it as it is?

Reply:
Maybe try and change the getter to: `public Optional<LocalDateTime> getTimeOfLastSuccessfulCallToAdmin() {...}`

If that leads too follow-up issues that are difficult to solve, I would suggest to keep it as it is for now.

Reply:
This was leading to follow-up issues, so it is better to keep it as it is for now